core/src/main/java/com/linecorp/armeria/internal/ConnectionLimitingHandler.java

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelOption;
+
+/**
+ * Limit the number of open connections to the configured value.
+ * {@link ConnectionLimitingHandler} instance would be set to {@link ServerBootstrap#handler(ChannelHandler)}.
+ */
+@Sharable
+public final class ConnectionLimitingHandler extends ChannelInboundHandlerAdapter {
+
+    private static final Logger logger = LoggerFactory.getLogger(ConnectionLimitingHandler.class);
+
+    private final int maxNumConnections;
+    private final AtomicInteger numConnections = new AtomicInteger();
+
+    private final AtomicBoolean droppedLogLock = new AtomicBoolean();
+    private final LongAdder numDroppedConnections = new LongAdder();
+
+    public ConnectionLimitingHandler(int maxNumConnections) {
+        this.maxNumConnections = validateMaxNumConnections(maxNumConnections);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        final Channel child = (Channel) msg;
+
+        int conn = numConnections.incrementAndGet();
+        if (conn > 0 && conn <= maxNumConnections) {
+            child.closeFuture().addListener(future -> numConnections.decrementAndGet());
+            super.channelRead(ctx, msg);
+        } else {
+            numConnections.decrementAndGet();
+
+            // Set linger option to 0 to reset channel.
+            child.config().setOption(ChannelOption.SO_LINGER, 0);
+            child.unsafe().closeForcibly();
+
+            numDroppedConnections.increment();
+
+            if (droppedLogLock.compareAndSet(false, true)) {
+                ctx.executor().schedule(this::writeNumDroppedConnectionsLog, 1, TimeUnit.SECONDS);
+            }
+        }
+    }
+
+    private void writeNumDroppedConnectionsLog() {
+        // There might be a missing count because resetting the counter and unlocking don't execute atomically.
+        // But it might be very few, and this is just a log message. So we ignore the missing count.
+        long dropped = numDroppedConnections.sumThenReset();
+        droppedLogLock.set(false);
```

Comment:
There's a race condition here:

1. `writeNumDroppedConnectionsLog()` is scheduled by `channelRead()`
2. `writeNumDroppedConnectionsLog()` resets the counter, but `droppedLogLock` is not set to false yet.
3. `channelRead()` tries `droppedLogLock.comparAndSet(false, true)` and it fails. `writeNumDroppedConnectionLog()` is not scheduled. `numDroppedConnections` becomes 1.
4. `writeNumDroppedConnectionsLog()` continues its execution till its end.
5. At this point, the lastly dropped connection will not be logged until another connection is dropped, which is incorrect.

A simple fix for this would be to:

- Set `droppedLogLock` to false before calling `sumThenReset()`, and
- Do not log when `dropped` is 0. (because of another race condition which can be triggered by setting `droppedLogLock` to false first.)